# Analysis of multiple fungal and bacterial sequence repositories highlights shortcomings in microbial databases

This project contains the links to the reference databases and the code that was used for our study : [Analysis of multiple fungal sequence repositories highlights shortcomings in microbial databases](https://www.biorxiv.org/content/10.1101/497867v2)

Currently, the study looks at the composition of the four  most widely used fungal genomic reference databases: 
 * [RefSeq](https://www.ncbi.nlm.nih.gov/refseq/)
 * [Ensembl](http://fungi.ensembl.org/info/website/ftp/index.html)
 * [FungiDB](https://fungidb.org/common/downloads/release-41/)
 * [JGI 1000 Fungal Genomes Project](https://genome.jgi.doe.gov/fungi/fungi.info.html)
 
... and three of the most widely used bacterial genomic reference databases: 
* [RefSeq](https://www.ncbi.nlm.nih.gov/refseq/)
* [Ensembl](http://bacteria.ensembl.org/info/website/ftp/index.html)
* [PATRIC](https://www.patricbrc.org/view/Taxonomy/2)

#### Table of Contents
##### Fungal Databases

* [Download Reference Database] ()
  * [RefSeq](https://github.com/smangul1/db.microbiome/tree/master/Fungi/code/Reference_Seq_DB_Download/NCBI_download)
  * [Ensmebl](https://github.com/smangul1/db.microbiome/tree/master/Fungi/code/Reference_Seq_DB_Download/Ensembl_download)
  * [FungiDB](https://github.com/smangul1/db.microbiome/tree/master/Fungi/code/Reference_Seq_DB_Download/FungiDB_download)
* [Jupyter notebook analysis](https://github.com/smangul1/db.microbiome/blob/master/Fungi/jupyter_notebook_analysis/analysis.ipynb)
* [SQL database](https://github.com/smangul1/db.microbiome/tree/master/Fungi/data)
* [Figure 1/2](https://github.com/smangul1/db.microbiome/tree/master/Fungi/jupyter_notebook_analysis/primary_fig)
* [Supplementary Figures](https://github.com/smangul1/db.microbiome/tree/master/Fungi/jupyter_notebook_analysis/supplementary_fig)

##### Bacterial Databases

* [Download Reference Database](https://github.com/Mangul-Lab-USC/db.microbiome/tree/master/Bacteria/Download_Databases)
  * [RefSeq]()
  * [Ensmebl](https://github.com/Mangul-Lab-USC/db.microbiome/blob/master/Bacteria/Download_Databases/Ensembl.sh)
  * [PATRIC](https://github.com/Mangul-Lab-USC/db.microbiome/blob/master/Bacteria/Download_Databases/PATRIC.sh)
* [Jupyter notebook analysis](https://github.com/Mangul-Lab-USC/db.microbiome/tree/master/Bacteria/jupyter_notebook_analysis)
* [SQL databases](https://github.com/Mangul-Lab-USC/db.microbiome/tree/master/Bacteria/data)
* [Figure 1/2](https://github.com/Mangul-Lab-USC/db.microbiome/tree/master/Bacteria/jupyter_notebook_analysis/primary_figures)
* [Supplementary Figures](https://github.com/Mangul-Lab-USC/db.microbiome/tree/master/Bacteria/jupyter_notebook_analysis/supplementary_figures)

# How to cite this study

> Loeffler, Caitlin, et al. "Improving the usability and comprehensiveness of microbial databases" bioRxiv, doi:[https://doi.org/10.1101/497867](https://doi.org/10.1101/497867) 
